pkg/destroy/aws/aws.go

```diff
@@ -77,1510 +63,1205 @@ func (o *ClusterUninstaller) validate() error {
-	funcs["deleteIAMresources"] = deleteIAMresources
-	funcs["deleteSecurityGroups"] = deleteSecurityGroups
-	funcs["deleteInternetGateways"] = deleteInternetGateways
-	funcs["deleteSubnets"] = deleteSubnets
-	funcs["deleteS3Buckets"] = deleteS3Buckets
-	funcs["deleteRoute53"] = deleteRoute53
-	funcs["deletePVs"] = deletePVs
-	funcs["deleteUsers"] = deleteUsers
-}
-
 // Run is the entrypoint to start the uninstall process
 func (o *ClusterUninstaller) Run() error {
 	err := o.validate()
 	if err != nil {
 		return err
 	}
 
-	deleteFuncs := map[string]deleteFunc{}
-	populateDeleteFuncs(deleteFuncs)
-	returnChannel := make(chan string)
+	awsConfig := &aws.Config{Region: aws.String(o.Region)}
 
-	awsSession, err := getAWSSession(o.Region)
+	// Relying on appropriate AWS ENV vars (eg AWS_PROFILE, AWS_ACCESS_KEY_ID, etc)
+	awsSession, err := session.NewSession(awsConfig)
 	if err != nil {
 		return err
 	}
 
-	// launch goroutines
-	goroutines := 0
-	for name, function := range deleteFuncs {
-		for _, filter := range o.Filters {
-			go deleteRunner(name, function, awsSession, filter, o.ClusterName, o.Logger, returnChannel)
-			goroutines++
+	tagClients := []*resourcegroupstaggingapi.ResourceGroupsTaggingAPI{
+		resourcegroupstaggingapi.New(awsSession),
+	}
+	tagClientNames := map[*resourcegroupstaggingapi.ResourceGroupsTaggingAPI]string{
+		tagClients[0]: o.Region,
+	}
+	if o.Region != "us-east-1" {
+		tagClient := resourcegroupstaggingapi.New(
+			awsSession, aws.NewConfig().WithRegion("us-east-1"),
+		)
+		tagClients = append(tagClients, tagClient)
+		tagClientNames[tagClient] = "us-east-1"
+	}
+
+	deleted := map[string]struct{}{}
+	iamClient := iam.New(awsSession)
+	iamRoleSearch := &iamRoleSearch{
+		client:  iamClient,
+		filters: o.Filters,
+		logger:  o.Logger,
+	}
+	iamUserSearch := &iamUserSearch{
+		client:  iamClient,
+		filters: o.Filters,
+		logger:  o.Logger,
+	}
+
+	var loopError error
+	for len(tagClients) > 0 || loopError != nil {
+		loopError = nil
+		nextTagClients := tagClients[:0]
+		for _, tagClient := range tagClients {
+			matched := false
+			for _, filter := range o.Filters {
+				o.Logger.Debugf("search for and delete matching resources by tag in %s matching %#+v", tagClientNames[tagClient], filter)
+				tagFilters := make([]*resourcegroupstaggingapi.TagFilter, 0, len(filter))
+				for key, value := range filter {
+					tagFilters = append(tagFilters, &resourcegroupstaggingapi.TagFilter{
+						Key:    aws.String(key),
+						Values: []*string{aws.String(value)},
+					})
+				}
+				err = tagClient.GetResourcesPages(
+					&resourcegroupstaggingapi.GetResourcesInput{TagFilters: tagFilters},
+					func(results *resourcegroupstaggingapi.GetResourcesOutput, lastPage bool) bool {
+						for _, resource := range results.ResourceTagMappingList {
+							arn := *resource.ResourceARN
+							if _, ok := deleted[arn]; !ok {
+								matched = true
+								err := deleteARN(awsSession, arn, o.Logger)
+								if err == nil {
```

Comment:
i think we can short-circuit to next item on non-nil error
```go
err := errors.Wrapf(deleteARN(awsSession, arn, o.Logger), "deleting %s", arn)
if err != nil {
    o.Logger.Debug(err)
    continue
}
deleted[arn] = exists
```

Reply:
although the current wrap on error is fine.

Reply:
> i think we can short-circuit to next item on non-nil error...

Done with 3a06697 -> 080152a.